src/transport/raw/TCP.cpp

```diff
@@ -314,75 +305,145 @@ CHIP_ERROR TCPBase::ProcessSingleMessageFromBufferHead(const PeerAddress & peerA
     return err;
 }
 
+CHIP_ERROR TCPBase::ActiveConnectionState::AckUnconsumed()
+{
+    // Ack the currently received data, so that we can be sent more.
+    uint16_t currentLength = mReceived->TotalLength();
+    if (mAckedNotConsumedLength > currentLength)
+    {
+        // This implies that some of mReceived has been consumed without calling AckConsumed().
+        return CHIP_ERROR_INTERNAL;
+    }
+    uint16_t needsAck = static_cast<uint16_t>(currentLength - mAckedNotConsumedLength);
+    if (needsAck != 0)
+    {
+        mAckedNotConsumedLength = currentLength;
+        return mEndPoint->AckReceive(needsAck);
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR TCPBase::ActiveConnectionState::AckConsumed(uint16_t length)
+{
+    if (mAckedNotConsumedLength != 0)
+    {
+        uint16_t ackedLength    = chip::min(mAckedNotConsumedLength, length);
```

Comment:
It looks like this code is trying to account for the case where `mAckedNotConsumedLength > length`.  However, in such a case, it will break.  For instance, if `mAckedNotConsumedLength == 110` and `length == 100`, the code will call `AckReceive(65526)`.

Suggest instead:

```
CHIP_ERROR TCPBase::ActiveConnectionState::AckConsumed(uint16_t length)
{
    if (mAckedNotConsumedLength > length)
    {
        mAckedNotConsumedLength = static_cast<uint16_t>(mAckedNotConsumedLength - length);
        length = 0;
    }
    else
    {
        length = static_cast<uint16_t>(length - mAckedNotConsumedLength);
        mAckedNotConsumedLength = 0;
    }
    if (length != 0)
    {
        return mEndPoint->AckReceive(length);
    }
    return CHIP_NO_ERROR;
}
```

Reply:
I may be fooling myself, but if `mAckedNotConsumedLength == 110` and `length == 100`,
```
uint16_t ackedLength    = chip::min(mAckedNotConsumedLength, length);                   ⇨ ackedLength := min(110,100) = 100
length                  = static_cast<uint16_t>(length - ackedLength);                  ⇨ length := 100 - 100 = 0
mAckedNotConsumedLength = static_cast<uint16_t>(mAckedNotConsumedLength - ackedLength); ⇨ 110 - 100 := 10
```
so there's no call to `AckReceive` and 10 bytes remaining acked but not consumed.

Reply:
Ah, yes you're right.  Thanks for explaining.